File: solution/minimum_spanning_tree/1922/main.py
# Authored by : gusdn3477
# Co-authored by : -
# Link : http://boj.kr/e8d47d936e104388916666605df5778d

import sys

def input():
    return sys.stdin.readline().rstrip()

def union(x, y):
    x = find_parent(x)
    y = find_parent(y)
    if x < y:
        parent[y] = x
    else:
        parent[x] = y

def find_parent(x):
    if x != parent[x]:
        parent[x] = find_parent(parent[x])
    return parent[x]
    
N = int(input())
M = int(input())
parent = [i for i in range(N+1)]
arr = []
ans = 0
for i in range(M):
    arr.append(list(map(int, input().split())))
arr.sort(key = lambda x : x[2])
for i in range(M):
    if find_parent(arr[i][0]) != find_parent(arr[i][1]):
        union(arr[i][0], arr[i][1])
        ans += arr[i][2]
print(ans)

File: solution/minimum_spanning_tree/1922/Main.java
//Authored by : suin8
//Co-authored by : -
//Link : http://boj.kr/7f20b69d2f604d2aaf5509387c9fb531

import java.util.*;
import java.io.*;

public class Main {
    static ArrayList<Edge> edge;
    static int p[];
    public static void main(String[] args) {
        FastReader rd = new FastReader();
        
        int N = rd.nextInt();
        int M = rd.nextInt();
        edge = new ArrayList<Edge>(N + 1);
        p = new int[N + 1];
        for(int i = 1;i <= N;i++)
            p[i] = i;
        
        for(int i = 0;i < M;i++) {
            int a = rd.nextInt();
            int b = rd.nextInt();
            int c = rd.nextInt();
            
            edge.add(new Edge(a, b, c));
        }
        // 오름차순 정렬
        Collections.sort(edge);
        
        int ans = 0;
        for(Edge e : edge) {
            // 사이클이 아닌 경우에만 포함시킨다
            if(find(e.n1) != find(e.n2)) {
                ans += e.weight;
                union(e.n1, e.n2);
            }
        }
        System.out.println(ans);
    }
    
    static int find(int num) {
        if(p[num] == num) return num;
        else return p[num] = find(p[num]);
    }
    
    static void union(int a, int b) {
        a = find(a);
        b = find(b);
        
        if(a < b) p[b] = a;
        else p[a] = b;
    }
    
    static class Edge implements Comparable<Edge> {
        int n1, n2, weight;
        
        Edge(int a, int b, int c) {
            this.n1 = a;
            this.n2 = b;
            this.weight = c;
        }
        
        @Override
        public int compareTo(Edge o) {
            return this.weight - o.weight;
        }
    }
    
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while(st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
        long nextLong() { return Long.parseLong(next()); }
        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}
